### ISSUE TITLE
Consistency Large Language Models: A Family of Efficient Parallel Decoders

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#11173]
Consistency Large Language Models: A Family of Efficient Parallel Decoders<br>
<br>
https://ift.tt/GFztZJl<br>
<br>
Siqi Kou*, Lanxiang Hu*, Zhezhi He, Zhijie Deng, Hao Zhang<br>
<br>
<div><div>
<p><strong>TL;DR:</strong> LLMs have been traditionally regarded as sequential decoders, decoding one token after another. In this blog, we show pretrained LLMs can be easily taught to operate as efficient parallel decoders. We introduce <strong>Consistency Large Language Models (CLLMs)</strong>, a new family of parallel decoders capable of reducing inference latency by efficiently decoding an -token sequence per inference step. Our research shows this process – mimicking human cognitive process of forming complete sentences in mind before articulating word by word – can be effectively learned by simply finetuning pretrained LLMs. Specifically, CLLMs are trained to perform parallel decoding by mapping any randomly initialized -token sequence to the same result yielded by autoregressive (AR) decoding in as few steps as possible. Experiment results show CLLMs obtained using our proposed method are highly effective, showing to improvements in generation speed, in par with or even beter than other fast inference techniques like Medusa2 and Eagle, yet require no additional memory cost to accomodate auxiliary model components at inference time.</p>
<figure><div>
<img alt="cllm-gsm8k-acc-demo" src="https://hao-ai-lab.github.io/blogs/cllm/img/baseline_vs_cllm_gsm8k_best_acc_demo.gif"><figcaption>Figure 1: Demo of speedup by CLLM-ABEL-7B-001 in comparison with baseline <a href="https://github.com/GAIR-NLP/abel">ABEL-7B-001</a> using Jacobi decoding on GSM8K.</figcaption>
</div></figure><h2>Background: Jacobi Decoding<a href="#background-jacobi-decoding">#</a>
</h2>
<p>Large language models (LLMs) are transforming the landscape of human lives, from programming to offering legal and health advice. However, during inference, LLMs generate responses token by token using AR decoding as shown in Figure 1, leading to high latency for longer responses. Using AR decoding, it often necessitates architectural modifications, auxiliary components, or draft models, to speed up inference by generating more than one token at a time.</p>
<figure><div>
<img alt="autoregressive" src="https://hao-ai-lab.github.io/blogs/cllm/img/clm_objective.png"><figcaption>Figure 2: illustration of conventional AR decoding: one token is generated at a time.</figcaption>
</div></figure><div>
<a href="https://arxiv.org/abs/2305.10427">Jacobi decoding</a><p> originates from the Jacobi and Gauss-Seidel fixed-point iteration for solving nonlinear equations, and </p>
<a href="https://proceedings.mlr.press/v139/song21a.html">is proven identical to AR generation using greedy decoding</a><p>. Jacobi decoding reformulates the sequential generation process into a system of </p>
<p> non-linear equations with </p>
<p> variables solvable in parallel based on Jacobi iteration. Each iteration step might predict more than one correct token (By “correct”, we mean alignment with the AR decoding
result under a greedy sampling strategy), thereby accelerating AR decoding potentially.</p>
</div>
<figure><div>
<img alt="jacobi" src="https://hao-ai-lab.github.io/blogs/cllm/img/jacobi_objective.png"><figcaption>Figure 3: illustration of Jacobi decoding: -token sequence is fed into the LLM and iterates until convergence.</figcaption>
</div></figure><p>To be specific, Jacobi decoding method first randomly guesses the next tokens in a sequence (referred to as -token sequence hereinafter unless specified otherwise) from an input prompt. The -token sequence, along with the prompt, is then fed to the LLM to iteratively update itself. This process continues until the -token sequence stabilizes and no further changes occur, reaching a fixed point. Notably, Jacobi decoding requires no more queries to the LLM than auto-regressive (AR) decoding. Eventually, the -token sequence converges to the output that would be generated by AR decoding under a greedy strategy. This progression from an initial random guess to the final AR generation outcome traces what is known as a Jacobi trajectory. An instance of Jacobi decoding iteration process and <strong>Jacobi trajectory</strong> is illustrated in Figure 2.</p>
<h3>Limitations of Jacobi Decoding<a href="#limitations-of-jacobi-decoding">#</a>
</h3>
<div>
<p>However, vanilla Jacobi decoding for LLMs shows only marginal speedup over AR decoding in practice, e.g., <a href="https://arxiv.org/abs/2305.10427">an average of speedup</a>. This is because an AR-trained LLM can rarely yield a correct token when there are incorrections in its preceding tokens. Thereby, most Jacobi iterations gain only one correction for the -token sequence, resulting in a longer trajectory as illustrated on the left side of Figure 3.</p>
<p><a href="https://lmsys.org/blog/2023-11-21-lookahead-decoding/">Lookahead decoding</a> and speculative decoding methods try to mitigate inefficiency in Jacobi decoding and conventional AR decoding, but incurs extra memory cost during inference time. While CLLMs require none.</p>
</div>
<h2>Consistency LLMs (CLLMs)<a href="#consistency-llms-cllms">#</a>
</h2>
<h3>Jacobi Decoding Preliminary<a href="#jacobi-decoding-preliminary">#</a>
</h3>
<p>Given a prompt and a pre-trained LLM , LLM typically generates with the standard AR decoding method under the greedy strategy, i.e.</p>
<p>Jacobi decoding re-frames the LLM inference process as solving a system of nonlinear equations to transform the decoding process into a parallelizable computation. Consider, , we can rewrite the above equation as a system of nonlinear equations:</p>
<p>Note that the process exits at some k such that and we define as the <strong>fixed point</strong>, and as the <strong>Jacobi trajectory</strong>.</p>
<h3>Training with Jacobi Trajectories<a href="#training-with-jacobi-trajectories">#</a>
</h3>
<div>
<p>To address this, we propose adapting pre-trained LLMs so that they can consistently map any point </p>
<p> on the Jacobi trajectory </p>
<p> to the fixed point </p>
<p>. Surprisingly, we find such an objective is analogous to that of </p>
<a href="https://arxiv.org/abs/2303.01469">consistency models</a><p>, a leading acceleration approach for diffusion models. In our proposed method, we use Jacobi trajectories collected from a target model to train the model with a loss that encourages single-step convergence during Jacobi iterations. For each target model </p>
<p> to be adapted as a CLLM, the training consists of two parts:</p>
</div>
<div><ul><li>
<strong>Jacobi trajectory preparation:</strong> for each prompt, we sequentially perform Jacobi decoding for every truncation of tokens until the entire response sequence has been generated, which amounts to a concatenation of all consecutive fixed points. Each sequence generated along a trajectory counts as one data entry. Note that for a lengthy response of () tokens, such truncation avoids slow model evaluation on lengthy input.</li></ul></div>
<div><ul><li>
<strong>Training with consistency and AR loss:</strong> we jointly optimize two losses for tuning CLLMs, the consistency loss guarantees the prediction of multiple tokens at once and the AR loss prevents the CLLM from deviating from the target LLM so as to maintain generation quality.</li></ul></div>
<figure><div>
<img alt="training_objective" src="https://hao-ai-lab.github.io/blogs/cllm/img/cllm_objective.png"><figcaption>Figure 4: an illustration of consistency training for one-step convergence: refining the target LLM to consistently predict the fixed point given any state along Jacobi trajectory as input.</figcaption>
</div></figure><h3>Consistency and AR Loss<a href="#consistency-and-ar-loss">#</a>
</h3>
<h4>Consistency Loss<a href="#consistency-loss">#</a>
</h4>
<div>
<p>Let denote the target LLM. Let denote the CLLM with parameters initialized with those of . For a prompt and the corresponding Jacobi trajectory , let and denote a random state and the fixed point on the trajectory, respectively.</p>
<p>We can encourage CLLM to output with as the input by minimizing the following loss, termed as the <strong>global consistency (GC) loss</strong>:</p>
</div>
<div>
<p>where and we abuse notations to represent uniform sampling from the dataset, and we abuse notations to represent uniform sampling from the dataset. denotes the distance between two distributions, choices are discussed in <a href="https://arxiv.org/abs/2306.13649">the GKD method</a> and in this paper we primarily experiment with the forward KL.</p>
<p>Alternatively, local consistency (LC) loss following the formulation in consistency models, where the adjacent states in a Jacobi trajectory are driven to yield the same outputs:</p>
</div>
<h4>AR Loss<a href="#ar-loss">#</a>
</h4>
<p>To avoid deviating from the distribution of the target LLM, we incorporate the traditional AR loss based on the generation of the target LLM :</p>
<p>Putting the two loss together, with some weight , the total loss for training a CLLM is:</p>
<h2>Experiments<a href="#experiments">#</a>
</h2>
<h3>Results<a href="#results">#</a>
</h3>
<p>Our experiments contain three domain-specific tasks, including Spider (text-to-SQL), Human-Eval (Python code completion), and GSM8k (math), and the broader open-domain conversational challenge, MT-bench. Reported experiments were conducted using either fine-tuned coder LLM, Deepseek-coder-7B-instruct, LLaMA-2-7B or ABEL-7B-001 as the target model depending on the task. Both training and evaluation are carried out on NVIDIA A100 40GB servers.</p>
<figure><div>
<img alt="speedup" src="https://hao-ai-lab.github.io/blogs/cllm/img/cllm_speedup.png"><figcaption>Figure 5: CLLM speedup on different downstream tasks. CLLMs are significantly faster than pre-trained models and achieve comparable speedups in comparison with Medusa, yet with no extra cost at inference time.</figcaption>
</div></figure><figure><div>
<img alt="specialized" src="https://hao-ai-lab.github.io/blogs/cllm/img/mt-bench.png">
<img alt="mt_bench" src="https://hao-ai-lab.github.io/blogs/cllm/img/specialized_domains.png">
</div>
<p><figcaption>Figure 6: illustration of CLLM vs. other baselines on domain-specific tasks (Spider, CSN-Python, GSM8k), as well as on MT-bench. CLLMs achieve similar or even better speedup in comoparison with Medusa2 while introducing no extra inference cost (in terms FLOPS and memory consumption).</figcaption></p></figure><div>
<p><strong>Specialized domains:</strong> From Figure 5, we can see that in comparison with other baselines including the original target model, Medusa2, and speculative decoding, CLLMs achieve the most significant speedup.</p>
<p><strong>Open-domain conversational Challenge (MT-bench):</strong> CLLM trained from LLaMA2-7B using ShareGPT dataset can achieve roughly the same speedup as Medusa2 when combined with lookahead decoding, with comparable scores on MT-bench. However, CLLM offers higher adaptability and memory efficiency as it requires no modifications to the target model’s original architecture and no auxiliary components.</p>
</div>
<h3>Training Cost<a href="#training-cost">#</a>
</h3>
<div>
<p>The fine-tuning cost of CLLMs is moderate, e.g., passing only around 1M tokens for LLaMA-7B to achieve a speedup on the Spider dataset. In the cases where the dataset size is large, for example, for CodeSearchNet-Python, only 10% of the dataset is required to generate Jacobi trajectories in training CLLMs to obtain around speedup. The total number of tokens can be estimated by taking:</p>
<p> avg # of trajectories per prompt avg trajectory length # of prompts.</p>
</div>
<div><table>
<thead><tr>
<th>dataset</th>
<th>estimated training cost (tokens)</th>
<th> of pre-training cost</th>
</tr></thead>
<tbody>
<tr>
<td>Spider</td>
<td>2M</td>
<td></td>
</tr>
<tr>
<td>CodeSearchNet-Python</td>
<td>100M</td>
<td></td>
</tr>
<tr>
<td>GSM8K</td>
<td>10M</td>
<td></td>
</tr>
<tr>
<td>ShareGPT</td>
<td>200M</td>
<td></td>
</tr>
</tbody>
</table></div>
<h3>Fast Forwarding and Stationary Tokens<a href="#fast-forwarding-and-stationary-tokens">#</a>
</h3>
<figure><div>
<img alt="trajectory_compare" src="https://hao-ai-lab.github.io/blogs/cllm/img/trajectory_compare_aligned.png"><figcaption>Figure 7: Comparison of Jacobi trajectory between a target LLM and CLLMs on Spider. Each point along the Jacobi trajectory is a color-coded sequence: blue for correct tokens matching with AR results, and red for inaccurate ones. CLLM demonstrates enhanced efficiency, converging to the fixed point faster the Target LLM. This increased efficiency in the CLLM can be attributed to the consistency loss which facilitates the learning of the structure of each -token sequence given a prefix.</figcaption>
</div></figure><div>
<p>The left side of Figure 6 shows target LLMs typically generate only one correct token in one iteration. In contrast, in CLLMs, we identify <strong>fast forwarding phenomenon</strong> where multiple consecutive tokens are correctly predicted in a single Jacobi iteration.</p>
<p>Moreover, tokens correctly generated in advance (e.g. “country” and “H” at index 6 and 7 on the left side of Figure 6), are often replaced inaccurately in subsequent iterations in target LLMs. On the other hand, CLLMs exhibit the capability of predicting correct tokens preemptively, even with preceding incorrect tokens, while ensuring the tokens remain unchanged. We term such tokens as <strong>stationary tokens</strong>. Both phenomena contribute to the fast convergence in Jacobi decoding of CLLMs, thereby leading to a considerable generation speedup.</p>
<p>We observe that CLLMs acquire a crucial linguistic concept through training – <strong>collocations</strong>: a series of words or terms that <a href="https://aclanthology.org/P91-1036.pdf">co-occur more frequently than one would expect by random chance</a>. Language is not solely composed of isolated words but also relies heavily on specific word pairings. Examples of collocations are abundant in both natural and coding languages. They include verb + preposition combinations (e.g., ‘’talk to’’, ‘‘remind … of …’’), verb + noun structures (e.g., ‘‘make a decision’’, ‘‘catch a cold’’), and many more domain-specific syntactical structures (e.g., ‘‘SELECT … FROM …’’, ‘‘if … else’’ for programming). The consistency generation objective allows CLLMs to infer such structures from any point in the Jacobi trajectory, encouraging CLLMs to acquire proficiency in numerous collocations and thereby predict multiple words simultaneously to minimize iteration steps.</p>
</div>
<h2>Get started<a href="#get-started">#</a>
</h2>
<h2>Acknowledgement<a href="#acknowledgement">#</a>
</h2>
<p>We would like to thank Yang Song, Canwen Xu, Yonghao Zhuang, Dacheng Li and Yichao Fu for providing insightful feedback.</p>
<h2>Citation<a href="#citation">#</a>
</h2>
<pre tabindex="0"><code>@misc{kou2024cllms,
      title={CLLMs: Consistency Large Language Models}, 
      author={Siqi Kou and Lanxiang Hu and Zhezhi He and Zhijie Deng and Hao Zhang},
      year={2024},
      eprint={2403.00835},
      archivePrefix={arXiv},
      primaryClass={cs.CL}
}
</code></pre>
</div></div><br>
<br>
<br>
<br>
<br>
<br>
via hao-ai-lab.github.io https://ift.tt/EvXedwO<br>
<br>
May 9, 2024 at 02:37PM
